data/kernels/color_conversion.cl

```diff
@@ -138,6 +139,16 @@ inline float get_rgb_matrix_luminance(const float4 rgb, constant dt_colorspaces_
   return luminance;
 }
 
+inline float4 rgb_matrix_to_xyz(const float4 rgb, constant dt_colorspaces_iccprofile_info_cl_t *profile_info, constant float *matrix, read_only image2d_t lut)
+{
+  if(profile_info->nonlinearlut)
+  {
+    float4 linear_rgb = apply_trc_in(rgb, profile_info, lut);
+    return matrix_product(linear_rgb, matrix);
+  }
+  return matrix_product(rgb, matrix);
```

Comment:
adding `return` in an `if` statement might prevent the compiler to optimize the branches (unswitch and reorder). In SIMD environments, the returned value should not depend on the branches or change the control-flow. Problem is here you return in branch so the control-flow may not predictable ahead.
so you can rewrite it:
```C
inline float4 rgb_matrix_to_xyz(const float4 rgb, constant dt_colorspaces_iccprofile_info_cl_t *profile_info, constant float *matrix, read_only image2d_t lut)
{
  float out = 0.f;
  if(profile_info->nonlinearlut)
  {
    float4 linear_rgb = apply_trc_in(rgb, profile_info, lut);
    out = matrix_product(linear_rgb, matrix);
  }
  else
  {
    out = matrix_product(rgb, matrix);
  }
  return out;
}
```